src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+        checkedId = checkedId,
+        startedAtSecond = startedAtSecond,
+        finishedAtSecond = finishedAtSecond,
+        roundIds = roundIds.toSortedSet,
+        decision = healtcheckDecision,
+        roundPeers = roundPeers,
+        parallelRounds = parallelRounds,
+        leftDuringConsensus = leftDuringConsensus,
+        removedUnresponsivePeersWithParallelRound = removedUnresponsivePeersWithParallelRound,
+        ownConsensusHealthStatus = ownConsensusHealthStatus,
+        isKeepUpRound = isKeepUpRound
+      )
+    } yield historicalRound
+
+  def processProposal(healthProposal: ConsensusHealthStatus): F[Unit] =
+    for {
+      _ <- roundIds.modify { roundIds =>
+        val updated = if (roundIds.contains(healthProposal.roundId)) roundIds else roundIds.add(healthProposal.roundId)
+
+        (updated, ())
+      }
+      initiallyUnknownPeers = healthProposal.clusterState.keySet
+        .diff(initialRoundPeers.keySet)
+        .filterNot(_ == ownId) // I think we don't need to filter out peers that left because they will not be added here by mistake
+      data <- roundPeers.modify { roundPeers =>
+        // observations are immutable
+        val peerRoundData = roundPeers
+          .get(healthProposal.checkingPeerId)
+          .map { roundData =>
+            val newConsensusData =
+              roundData.consensusData.getOrElse(
+                PeerConsensusData(healthProposal.status, healthProposal.clusterState, healthProposal.roundId)
+              )
+
+            roundData.copy(consensusData = newConsensusData.some)
+          }
+
+        val updated =
+          peerRoundData.map { prd =>
+            val updatedPeers = prd.peerData match {
+              case Left(_) =>
+                roundPeers //for now we don't allow proxy through more than one node - so we don't proxy p1 -> p2 -> p3 -> p4, because p1 sees p2 which sees p3 which sees p4
+              case Right(peerData) =>
+                val addedPeers =
+                  (initiallyUnknownPeers -- roundPeers.keySet)
+                    .map(
+                      _ -> RoundData(ProxyPeers(NonEmptySet.one(peerData)).asLeft[PeerData])
+                    )
+                val updatedProxyPeers =
+                  roundPeers
+                    .filterKeys(initiallyUnknownPeers.contains)
+                    .mapValues(
+                      roundData =>
+                        roundData.copy(
+                          peerData = roundData.peerData
+                            .leftMap(proxyPeers => proxyPeers.copy(proxyPeers.peersData.add(peerData)))
+                        )
+                    )
+
+                roundPeers ++ updatedProxyPeers ++ addedPeers
+            }
+
+            updatedPeers + (healthProposal.checkingPeerId -> prd)
+          }.getOrElse(roundPeers)
+        (updated, updated)
+      }
+    } yield ()
+
+  private def fetchProposals(): F[Unit] =
+    for {
+      isAcquired <- fetchingProposalLock.tryAcquire
```

Comment:
have you handled looping it in another place? afaik this will try once only 🤔

Reply:
And it's fine I think because the next time the management flow starts it will try again. So yes it's retried if not running with every periodic start.